File: c++/p005.cpp
#include <iostream>
#include <vector>
#include <set>
#include <cmath>
using namespace std;
vector<int> prime_factorization(int n) {
    vector<int> prime_factors;
    int c = 2; 
    prime_factors.push_back(c);
    while (n>1){
        if (n % c == 0){
            n = n / c;
            prime_factors.push_back(c);
        }
        else{
            c = c + 1;
        }
    }
    return prime_factors;   
}
vector<int> lcm(vector<int> nlist) {
    set<int> primes;
    for (size_t i = 0; i < nlist.size(); ++i) {
        vector<int> n = prime_factorization(nlist[i]);
        for (size_t j = 0; j < n.size(); ++j) {
            primes.insert(n[j]);
            }
        }
    vector<int>ans(primes.begin(),primes.end());
    return ans;
}

int compute(int k, vector<int>& p) {
    int N=1, i=0;
    bool check = true;
    double lim = sqrt(k);
    vector<int> a(p.size(),0);
    while (p[i] < k){
        a[i] = 1;
        if (check){
            if (p[i] <= lim){
                a[i] = floor(log(k) / log(p[i]));
            }
            else{
                check = false;
            }
        }
        N = N * pow(p[i],a[i]);
        i = i + 1;
        if(i==p.size()){
            break;
        }
    }
    return N;
}

int main() {
    vector<int> nlist;
    for (int i = 2; i <= 20; ++i) {
        nlist.push_back(i);
    }
    vector<int> factors = lcm(nlist);
    for (int i = 0; i < factors.size(); i++){
        cout<<factors[i]<<' ';
    }
    cout << '\n';
    int ans = compute(20,factors);
    cout << ans << endl;
    return 0;
}

File: python/p005.py
import numpy as np
import pandas as pd
import math
def prime_factorization(n):
    prime_factors = []
    c = 2 # smallest prime number
    while n > 1:
        if n % c == 0:
            n = n / c
            prime_factors.append(c)
        else:
            c = c + 1
    n,m = np.unique(prime_factors,return_counts=True)
    return n,m
def lcm(nlist,primes_val=True):
    primes = []
    powers = []
    for i in range(len(nlist)): 
        n,m = prime_factorization(nlist[i])
        if len(n)>1:
            for j in range(len(n)):
                primes.append(n[j])
                powers.append(m[j])
        else:
            primes.append(n[0])
            powers.append(m[0])
    df = pd.DataFrame(columns=['primes','powers'])
    df.loc[:,'primes'] = primes
    df.loc[:,'powers'] = powers
    primes_df = df.iloc[df.groupby(['primes']).apply(lambda x: x['powers'].idxmax())] 
    primes_df.loc[:,'mult'] = primes_df['primes'].values ** primes_df['powers'].values
    primes_df = primes_df.reset_index(drop=False)
    ans = primes_df['mult'].cumprod()
    if primes_val:
        return primes_df['primes'].values
    else:
        return ans.iloc[-1]

def compute(k,p):
    N,i = 1,0
    check = True
    lim = math.sqrt(k)
    a = np.zeros_like(p)
    while p[i] < k:
        #print (p[i-1])
        a[i] = 1
        if check:
            if p[i] <= lim:
                a[i] = math.floor(math.log(k) / math.log(p[i]))
            else:
                check = False
        N = N * p[i]**a[i]
        i = i+1
        if i==len(p):
            break
    return N 


if __name__ == "__main__":
    print ('Project Euler 5')
    nlist = list(range(2,21))
    n = lcm(nlist)
    print (n) 
    print (compute(20,[2,3,5,7,11,13,17,19]))

    